### ISSUE TITLE
Missing action hooks.

### ISSUE THREAD

**Issue by esbanarango** [sethbrasile/ember-remodal#23]
There are certain situations where we have to perform some logic before opening the modal. An action hook like `beforeOpen` would be __really__ handy to put this kind of logic; additionally, we could prevent the opening of the modal in this hook if for example we have a validation there.

**Comment by sethbrasile** [comment]
Wouldn't it be just as easy to perform whatever logic you would put there, inside the action you're using to open the modal? What would be the difference?

```
openModal() {
  this.get('remodal').open('modal1')
},
beforeOpenModal1() {
  // something that returns a promise
}
```
Would be the same as:
```
openModal() {
  somethingThatReturnsaPromise.then(() => {
    this.get('remodal').open('modal1');
  });
}
```

wouldn't it?

**Comment by esbanarango** [comment]
Yeah, that solution works when you’re opening the modal programmatically, but the case I’m facing is opening the modal using the `openButton`.

**Comment by sethbrasile** [comment]
That's a good point. I'm inclined to lean toward saying "that use-case calls for programmatically opening a modal", because I had envisioned the `openButton`'s purpose as only being for simple "click and a modal opens" situations, but I can see where allowing that to be conditional (based on some validation or something) could be a good thing. It certain situations it could lead to cleaner implementation in the consuming app. I'm going to leave this one open and think about it a bit.

If anyone else has input, please chime in!

**Comment by sethbrasile** [comment]
Also, I'm assuming if we did this we would also add hooks for `beforeCancel`, `beforeConfirm`, and `beforeClose`.

**Comment by esbanarango** [comment]
Agree! We will also need to add those.
